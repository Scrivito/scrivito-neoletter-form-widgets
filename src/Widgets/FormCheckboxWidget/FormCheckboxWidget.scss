.scrivito-neoletter-form-widgets.form-container-widget .form-check {
  .form-check-label {
    display: inline;
    
    p {
      display: inline;
      margin: 0;
    }
  }
}

