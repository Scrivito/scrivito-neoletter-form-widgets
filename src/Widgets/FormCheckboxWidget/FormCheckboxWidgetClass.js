import * as Scrivito from "scrivito";
import { getScrivitoFormWidgetConfig } from "../../config/scrivitoConfig";
export const FormCheckboxWidget = Scrivito.provideWidgetClass(
  "FormCheckboxWidget",
  {
    attributes: {
      type: [
        "enum",
        {
          values: ["custom", "accept_terms"].concat(
            getScrivitoFormWidgetConfig().neoletterSubmissionEnabled
              ? ["subscription"]
              : []
          ),
        },
      ],
      customFieldName: "string",
      label: "string",
      required: "boolean",
      helpText: "html",
    },
  }
);
