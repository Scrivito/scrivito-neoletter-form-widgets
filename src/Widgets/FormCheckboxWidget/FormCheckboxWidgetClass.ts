import * as Scrivito from "scrivito";
export const FormCheckboxWidget = Scrivito.provideWidgetClass(
  "FormCheckboxWidget",
  {
    attributes: {
      type: [
        "enum",
        {
          values: ["custom", "subscription", "accept_terms"]
        }
      ],
      customFieldName: "string",
      label: "string",
      required: "boolean",
      helpText: "html"
    }
  }
);
