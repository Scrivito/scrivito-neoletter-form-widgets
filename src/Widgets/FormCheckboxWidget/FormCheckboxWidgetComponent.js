import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";

Scrivito.provideComponent("FormCheckboxWidget", ({ widget }) => {
  const id = `form_checkbox_widget_${widget.id()}`;
  const labelOptions = {};
  if (!Scrivito.isInPlaceEditingActive()) {
    labelOptions.htmlFor = id;
  }

  return (
    <div className="form-check mb-3">
      <input
        className="form-check-input"
        id={id}
        type="checkbox"
        name={getFieldName(widget)}
        required={widget.get("required")}
      />

      <Scrivito.ContentTag
        className="form-check-label"
        content={widget}
        attribute="label"
        tag="label"
        {...labelOptions}
      />
      {widget.get("required") && <Mandatory />}
      {widget.get("helpText") && <HelpText widget={widget} />}
    </div>
  );
});
