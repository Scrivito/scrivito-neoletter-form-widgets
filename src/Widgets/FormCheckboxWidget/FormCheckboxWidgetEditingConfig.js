import * as Scrivito from "scrivito";
import formCheckboxWidgetIcon from "../../assets/images/form_checkbox_widget.svg";
import { isCustomType } from "../FormContainerWidget/utils/isCustomType";
import { customFieldNameValidation } from "../FormContainerWidget/utils/validations/customFieldNameValidation";
import { typeValidation } from "../FormContainerWidget/utils/validations/typeValidation";
import { insideFormContainerValidation } from "../FormContainerWidget/utils/validations/insideFormContainerValidation";
import { getScrivitoFormWidgetConfig } from "../../config/scrivitoConfig";
Scrivito.provideEditingConfig("FormCheckboxWidget", {
  title: "Form Checkbox",
  thumbnail: formCheckboxWidgetIcon,
  attributes: {
    required: { title: "Mandatory" },
    type: {
      title: "Input type",
      values: [
        { value: "accept_terms", title: "Accept terms" },
        ...(getScrivitoFormWidgetConfig().neoletterSubmissionEnabled
          ? [{ value: "subscription", title: "Subscription" }]
          : []),
        { value: "custom", title: "Custom" },
      ],
    },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" },
  },
  properties: (widget) =>
    isCustomType(widget)
      ? ["type", "customFieldName", "label", "required", "helpText"]
      : ["type", "label", "required", "helpText"],
  initialContent: {
    type: "custom",
    customFieldName: "custom_checkbox",
    label: "Please send me your free printed product catalog.",
  },
  validations: [
    insideFormContainerValidation,
    typeValidation,
    customFieldNameValidation,
  ],
});
