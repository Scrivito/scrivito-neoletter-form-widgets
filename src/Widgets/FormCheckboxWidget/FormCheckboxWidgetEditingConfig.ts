import * as Scrivito from "scrivito";
import formCheckboxWidgetIcon from "../../assets/images/form_checkbox_widget.svg";
import { isCustomType } from "../FormStepContainerWidget/utils/isCustomType";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { typeValidation } from "../FormStepContainerWidget/utils/validations/typeValidation";
import { insideFormContainerValidation } from "../FormStepContainerWidget/utils/validations/insideFormContainerValidation";
Scrivito.provideEditingConfig("FormCheckboxWidget", {
  title: "Neoletter Form Checkbox",
  thumbnail: formCheckboxWidgetIcon,
  attributes: {
    required: { title: "Mandatory" },
    type: {
      title: "Input type",
      values: [
        { value: "accept_terms", title: "Accept terms" },
        { value: "subscription", title: "Subscription" },
        { value: "custom", title: "Custom" }
      ]
    },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" }
  },
  properties: (widget: Scrivito.Widget) =>
    isCustomType(widget)
      ? ["type", "customFieldName", "label", "required", "helpText"]
      : ["type", "label", "required", "helpText"],
  initialContent: {
    type: "custom",
    customFieldName: "custom_checkbox",
    label: "Please send me your free printed product catalog."
  },
  validations: [
    insideFormContainerValidation,
    typeValidation,
    customFieldNameValidation
  ]
});
