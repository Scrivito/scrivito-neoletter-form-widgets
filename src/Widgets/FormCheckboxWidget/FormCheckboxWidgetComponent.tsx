import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { FormCheckboxWidget } from "./FormCheckboxWidgetClass";

Scrivito.provideComponent(FormCheckboxWidget, ({ widget }) => {
  const id = `form_checkbox_widget_${widget.id()}`;

  return (
    <div className="form-check mb-3">
      <input
        className="form-check-input"
        id={id}
        type="checkbox"
        name={getFieldName(widget)}
        required={widget.get("required")}
      />

      <Scrivito.ContentTag
        className="form-check-label"
        content={widget}
        attribute="label"
        tag="label"
        htmlFor={id}
      />
      {widget.get("required") && <Mandatory />}
      {widget.get("helpText") && <HelpText widget={widget} />}
    </div>
  );
});
