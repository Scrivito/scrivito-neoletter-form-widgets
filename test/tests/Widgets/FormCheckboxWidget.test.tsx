import * as Scrivito from "scrivito";
import { screen } from "@testing-library/react";
import { FormCheckboxWidget } from "../../../src/Widgets/FormCheckboxWidget/FormCheckboxWidgetClass";
import "../../../src/Widgets/FormCheckboxWidget/FormCheckboxWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

const widgetProps = {
  label: "Test Checkbox",
  required: true,
  helpText: "This is a test help text",
  type: "custom",
  customFieldName: "custom_test"
};

describe("FormCheckboxWidget", () => {
  it("displays the checkbox label, Mandatory component, and HelpText", () => {
    pageRenderer.render({
      body: [new FormCheckboxWidget(widgetProps)]
    });

    const checkboxLabel = screen.getByText(widgetProps.label);
    const mandatoryText = screen.getByText("*", {
      selector: ".text-mandatory"
    });
    const helpTextElement = document.querySelector("i");

    expect(checkboxLabel).toBeInTheDocument();
    expect(mandatoryText).toBeInTheDocument();
    expect(helpTextElement).toBeInTheDocument();
  });

  it("renders the checkbox input field correctly", () => {
    pageRenderer.render({
      body: [new FormCheckboxWidget(widgetProps)]
    });

    const checkboxInput = screen.getByRole("checkbox");

    expect(checkboxInput).toBeInTheDocument();
    expect(checkboxInput).toHaveAttribute("name", "custom_test");
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [new FormCheckboxWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
