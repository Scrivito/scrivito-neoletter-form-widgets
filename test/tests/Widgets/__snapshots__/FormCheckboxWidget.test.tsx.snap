// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="form-check mb-3"
    >
      <input
        className="form-check-input"
        id="form_checkbox_widget_staticWidgetID"
        name="custom_test"
        required={true}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="form_checkbox_widget_staticWidgetID"
      >
        Test Checkbox
      </label>
      <div
        className="mandatory-container"
      >
        <span
          className="text-mandatory"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          *
        </span>
      </div>
      <div
        className="helptext-container"
      >
        <i
          className="bi bi-question-circle"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
